restapi/src/main/java/io/stargate/web/docsapi/service/query/search/resolver/impl/CnfResolver.java

```diff
@@ -0,0 +1,247 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service.query.search.resolver.impl;
+
+import com.bpodgursky.jbool_expressions.And;
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Literal;
+import com.bpodgursky.jbool_expressions.options.ExprOptions;
+import com.bpodgursky.jbool_expressions.rules.Rule;
+import com.bpodgursky.jbool_expressions.rules.RuleList;
+import com.bpodgursky.jbool_expressions.rules.RulesHelper;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import io.reactivex.rxjava3.core.Flowable;
+import io.stargate.web.docsapi.dao.DocumentDB;
+import io.stargate.web.docsapi.dao.Paginator;
+import io.stargate.web.docsapi.service.DocsApiConfiguration;
+import io.stargate.web.docsapi.service.ExecutionContext;
+import io.stargate.web.docsapi.service.RawDocument;
+import io.stargate.web.docsapi.service.query.FilterExpression;
+import io.stargate.web.docsapi.service.query.FilterPath;
+import io.stargate.web.docsapi.service.query.rules.TrueFilterExpressions;
+import io.stargate.web.docsapi.service.query.search.resolver.BaseResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.DocumentsResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.CandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.impl.InMemoryCandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.impl.PersistenceCandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.weigth.ExpressionWeightResolver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/** Resolver that accepts the {@link And} expression in the CNF form. */
+public class CnfResolver implements DocumentsResolver {
+
+  private final DocumentsResolver delegate;
+
+  public CnfResolver(
+      Expression<FilterExpression> expression,
+      ExpressionWeightResolver<FilterExpression> weightResolver,
+      ExecutionContext context,
+      DocumentsResolver parent) {
+    if (!And.EXPR_TYPE.equals(expression.getExprType())) {
+      throw new IllegalArgumentException(
+          "CNF resolver expect the AND expression in the conjunctive normal form.");
+    }
+
+    this.delegate = resolveInternal(expression, weightResolver, context, parent);
```

Comment:
Not sure what we gain and it seems like again a lot of time to do it like so.. I see a benefit of having resolvers mirror the epxression tree in debugging, don't you as well?

Reply:
Absolutely, `Resolver` implementations will mirror the query plan structure. However, `delegate` looks a bit foreign in this context. I think we can resolve this delegation when we construct document resolvers so that all final `Resolver` implementation will always to "real" work in `getDocuments` rather than call `delegate`. WDYT?

Reply:
moved to static for cnf and single resolver